### ISSUE TITLE
gtkhello sample does not work on macOS

### ISSUE THREAD

**Issue by izarif** [lgi-devs/lgi#151]
Terminal output:
```
lua-5.3: /usr/local/share/lua/5.3/lgi/override/cairo.lua:31: attempt to index a nil value (field '_module')
stack traceback:
	/usr/local/share/lua/5.3/lgi/override/cairo.lua:31: in main chunk
	[C]: in function 'require'
	/usr/local/share/lua/5.3/lgi/namespace.lua:176: in function 'lgi.require'
	/usr/local/share/lua/5.3/lgi/namespace.lua:163: in function 'lgi.require'
	/usr/local/share/lua/5.3/lgi/namespace.lua:163: in function 'lgi.require'
	create-main-window.lua:4: in main chunk
	[C]: in ?
```

**Comment by psychon** [comment]
Don't ask me how to get it, but I think this means that you do not have the GObject introspection typelib for cairo.

Can you run `lua -e 'print(require("lgi.core").module("cairo", 2))' `? It should output something like `userdata: 0x5633a73ef358	libcairo.so.2`. I guess in your case the first value will be `nil`.

@pavouk If the above is correct, I'll leave it to you to fix the code so that a better error message is generated when cairo is not available. Also, I don't know which other overrides might be affected.

**Comment by TingPing** [comment]
> Don't ask me how to get it, but I think this means that you do not have the GObject introspection typelib for cairo.

It is generated at build time of `gobject-introspection` if `cairo` **and** `cairo-gobject` libraries are found.

**Comment by speeder** [comment]
I am also having this problem, I tried copy-pasting that print line above, this is the result:

"nil	/usr/local/Cellar/gobject-introspection/1.50.0/lib/libcairo.2.dylib"

I have no idea what is wrong.

**Comment by ffauberteau** [comment]
I obtain the same error on NetBSD:
`Lua 5.3.4  Copyright (C) 1994-2017 Lua.org, PUC-Rio`
`> lgi = require("lgi")`
`> cairo = lgi.cairo`
`/usr/pkg/share/lua/5.3/lgi/override/cairo.lua:31: attempt to index a nil value (field '_module')`
`stack traceback:`
`        /usr/pkg/share/lua/5.3/lgi/override/cairo.lua:31: in main chunk`
`        [C]: in function 'require'`
`        /usr/pkg/share/lua/5.3/lgi/namespace.lua:176: in function 'lgi.namespace.require'`
`        (...tail calls...)`
`        stdin:1: in main chunk`
`        [C]: in ?`
`> print(require("lgi.core")).module("cairo",2)`
`table: 0x7f7ff7b12100`
`stdin:1: attempt to index a nil value`
`stack traceback:`
`        stdin:1: in main chunk`
`        [C]: in ?`

However, typelib for cairo is provided by gobject-introspection:
`pkg_info -L gobject-introspection|grep cairo`
`/usr/pkg/lib/girepository-1.0/cairo-1.0.typelib`

I would be very happy to receive any advice to identify the problem.

**Comment by psychon** [comment]
@ffauberteau Can you try `print(require("lgi.core").module("cairo",2))` instead of `print(require("lgi.core")).module("cairo",2)`? (`print` around everything, not just the `require`)

**Comment by ffauberteau** [comment]
@psychon You're right, I write a typo. But my first error was correct.
The problem was a building dependency. Now it works.
Thanks

**Comment by psychon** [comment]
@ffauberteau Could you provide more information about that dependency problem? Could others have the same problem and your solution would work for them as well?

**Comment by andrewstarks** [comment]
I'm having this same issue with lgi. Can @ffauberteau or anyone else shed a bit of light on what the fix was?

**Comment by jorge-jbs** [comment]
Same issue here. I'm trying to use this library with the nix package manager and I can't make it work. Help would be appreciated.

**Comment by psychon** [comment]
@jorge-jbs "Nix" sounds a bit like "Linux" to me, so what does `lua -e 'print(require("lgi.core").module("cairo", 2))'` say for you? Also, no idea about the correct path for it (debian uses `/usr/lib/x86_64-linux-gnu/libcairo.so.2` here), but where is your cairo library & is it called `libcairo.so.2`?

**Comment by jorge-jbs** [comment]
@psychon It prints `nil	libcairo.so.2`. Nix is not an usual package manager, cairo is located in `/nix/store/mwmwbrnj2qvfcwscqx1fxfzzlhqfp011-cairo-1.14.10/lib/libcairo.so.2`. But I don't think Nix is the problem.

**Comment by psychon** [comment]
@jorge-jbs Hm, so it did not find `libcairo.so.2`. LGI just uses [`g_module_open("libcairo.so.2", 0)`](https://developer.gnome.org/glib/stable/glib-Dynamic-Loading-of-Modules.html#g-module-open). In this case, it looks to me like the code in there just does `dlopen("libcairo.so.2", RTLD_NOW)`. So all that I can tell you is that the dynamic loader does not find libcairo on your system / that the file is not in its search path.
Code for `g_module_open`: https://sources.debian.net/src/glib2.0/2.53.4-3/gmodule/gmodule.c/#L595 https://sources.debian.net/src/glib2.0/2.53.4-3/gmodule/gmodule-dl.c/#L98

**Comment by psychon** [comment]
@jorge-jbs Google found the following and to me it sounds like `dlopen` doesn't have much of a search path on Nix. You could try running with `LD_LIBRARY_PATH=/nix/store/mwmwbrnj2qvfcwscqx1fxfzzlhqfp011-cairo-1.14.10/lib/` I guess, but that's mostly a hack. I don't see why you think that

> But I don't think Nix is the problem.

https://nixos.org/nix-dev/2011-December/007412.html
https://nixos.org/nix-dev/2014-August/013822.html

**Comment by medranocalvo** [comment]
It seems to me that the change in https://github.com/pavouk/lgi/commit/dab172c9769e161037e59bc34f9a665d33b89736 prevents the library search path from being used: an absolute path is used unconditionally in macOS.  I think this commit should be reverted. Instead, users/system administrators should make sure the libraries are available in the library search path.  It would be helpful for the error message to include a comment about setting LD_LIBRARY_PATH, or using pkg-config to find out what the library search path is.

**Comment by andrewstarks** [comment]
I've been following this thread, googling and otherwise trying to solve this problem on my Mac (Mojave) and i cannot figure it out. 

I have installed:  Lua 5.3, Lgi (latest), gtk+, gtk+3, gobject-introspection, gstreamer, gst-plugins-bad/base/good/ugly, and cairo

For clarity in debugging, I've patched override/cairo.lua like so:
```Lua
-- <snip>
local err --new
cairo._module, err = core.module('cairo', 2)
---[[ New
if not cairo._module then
     error(string.format(
          "_module not loaded. Error message: %s", err or '<nil>'))
end 
--]]
```
and get the following output:
```
> lgi = require'lgi'
> cairo = lgi.cairo
/usr/local/share/lua/5.3/lgi/override/cairo.lua:26: _module not loaded. Error message: /usr/local/Cellar/gobject-introspection/1.54.1/lib/libcairo.2.dylib
stack traceback:
	[C]: in function 'error'
	/usr/local/share/lua/5.3/lgi/override/cairo.lua:26: in main chunk
	[C]: in function 'require'
	/usr/local/share/lua/5.3/lgi/namespace.lua:183: in function 'lgi.require'
	(...tail calls...)
	stdin:1: in main chunk
	[C]: in ?
> 
```

Poking around on my hard drive reveals this:

```Zsh
andrew:~/ $ pkg-config cairo --libs                                                                            [14:54:53]
-L/usr/local/Cellar/cairo/HEAD-ef8c379/lib -lcairo
andrew:~/ $ ls /usr/local/Cellar/cairo/HEAD-ef8c379/lib                                                        [14:55:01]
total 5264
drwxr-xr-x  13 andrew  admin      416 Jan 23 11:12 .
drwxr-xr-x  12 andrew  admin      384 Jan 23 11:12 ..
drwxr-xr-x   8 andrew  admin      256 Jan 23 11:12 cairo
-r--r--r--   1 andrew  admin    45084 Jan 23 11:12 libcairo-gobject.2.dylib
-r--r--r--   1 andrew  admin    31536 Jan 23 11:11 libcairo-gobject.a
lrwxr-xr-x   1 andrew  admin       24 Jan 23 11:11 libcairo-gobject.dylib -> libcairo-gobject.2.dylib
-r--r--r--   1 andrew  admin   129716 Jan 23 11:12 libcairo-script-interpreter.2.dylib
-r--r--r--   1 andrew  admin   151608 Jan 23 11:12 libcairo-script-interpreter.a
lrwxr-xr-x   1 andrew  admin       35 Jan 23 11:12 libcairo-script-interpreter.dylib -> libcairo-script-interpreter.2.dylib
-r--r--r--   1 andrew  admin   884088 Jan 23 11:12 libcairo.2.dylib
-r--r--r--   1 andrew  admin  1438336 Jan 23 11:11 libcairo.a
lrwxr-xr-x   1 andrew  admin       16 Jan 23 11:11 libcairo.dylib -> libcairo.2.dylib
drwxr-xr-x  15 andrew  admin      480 Jan 23 11:11 pkgconfig
andrew:~/ $                                                         
```
Other tutorials work, as long as they do not make use of Cairo.


Does anyone have a suggestion on what I might try next.

**Comment by psychon** [comment]
@andrewstarks So, lgi looks in ` /usr/local/Cellar/gobject-introspection/1.54.1/lib/libcairo.2.dylib` (the library dir of gobject-introspection), but your library is actually in `/usr/local/Cellar/cairo/HEAD-ef8c379/lib/libcairo.2.dylib` (the path where Cellar put it). I'd suggest to just create a symbolic link to make this path actually work.
```sh
ln -s /usr/local/Cellar/cairo/HEAD-ef8c379/lib/libcairo.2.dylib  /usr/local/Cellar/gobject-introspection/1.54.1/lib/libcairo.2.dylib
```

**Comment by andrewstarks** [comment]
Thank you for that advice! It is much appreciated. I tried your suggestion.

Problem 1:

I don't actually have that version installed I have 1.58.3 installed. If I adjust your command and make the link, it still doesn't find it (of course). 

Any clues on where the string '1.54.1' may be lingering? Maybe an old installation, but I've tried uninstalling and reinstalling just about everything (lgi, gtk+, all of the homebrew installed files).

How do I change gointrospection to look in the right spot? (sorry if I'm asking this in the wrong place)

Problem 2?:

My directory structure looks different than that. Doing digging finds: 

```
andrew:~/ $ ls /usr/local/Cellar/gobject-introspection/1.58.3/lib                                                                                  
total 976
drwxr-xr-x   8 andrew  admin     256 Jan 23 11:17 .
drwxr-xr-x  12 andrew  admin     384 Jan 23 11:17 ..
drwxr-xr-x  19 andrew  admin     608 Jan 23 11:17 girepository-1.0
drwxr-xr-x   3 andrew  admin      96 Jan 23 11:17 gobject-introspection
-r--r--r--   1 andrew  admin  187428 Jan 23 11:17 libgirepository-1.0.1.dylib
-r--r--r--   1 andrew  admin  311160 Jan 23 11:17 libgirepository-1.0.a
lrwxr-xr-x   1 andrew  admin      27 Jan 23 11:17 libgirepository-1.0.dylib -> libgirepository-1.0.1.dylib
drwxr-xr-x   4 andrew  admin     128 Jan 23 11:17 pkgconfig
andrew:~/ $ ls /usr/local/Cellar/gobject-introspection/1.58.3/lib/gobject-introspection                                                                                               
total 0
drwxr-xr-x   3 andrew  admin    96 Jan 23 11:17 .
drwxr-xr-x   8 andrew  admin   256 Jan 23 11:17 ..
drwxr-xr-x  35 andrew  admin  1120 Jan 23 11:17 giscanner
andrew:~/ $ ls /usr/local/Cellar/gobject-introspection/1.58.3/lib/girepository-1.0                                
total 1360
drwxr-xr-x  19 andrew  admin     608 Jan 23 11:17 .
drwxr-xr-x   8 andrew  admin     256 Jan 23 11:17 ..
-r--r--r--   1 andrew  admin     712 Jan 23 11:17 DBus-1.0.typelib
-r--r--r--   1 andrew  admin     560 Jan 23 11:17 DBusGLib-1.0.typelib
-r--r--r--   1 andrew  admin   26840 Jan 23 11:17 GIRepository-2.0.typelib
-r--r--r--   1 andrew  admin     948 Jan 23 11:17 GL-1.0.typelib
-r--r--r--   1 andrew  admin  190392 Jan 23 11:17 GLib-2.0.typelib
-r--r--r--   1 andrew  admin    1368 Jan 23 11:17 GModule-2.0.typelib
-r--r--r--   1 andrew  admin   59044 Jan 23 11:17 GObject-2.0.typelib
-r--r--r--   1 andrew  admin  344796 Jan 23 11:17 Gio-2.0.typelib
-r--r--r--   1 andrew  admin   14172 Jan 23 11:17 cairo-1.0.typelib
-r--r--r--   1 andrew  admin     348 Jan 23 11:17 fontconfig-2.0.typelib
-r--r--r--   1 andrew  admin     420 Jan 23 11:17 freetype2-2.0.typelib
-r--r--r--   1 andrew  admin     668 Jan 23 11:17 libxml2-2.0.typelib
-r--r--r--   1 andrew  admin     176 Jan 23 11:17 win32-1.0.typelib
-r--r--r--   1 andrew  admin     240 Jan 23 11:17 xfixes-4.0.typelib
-r--r--r--   1 andrew  admin     464 Jan 23 11:17 xft-2.0.typelib
-r--r--r--   1 andrew  admin     836 Jan 23 11:17 xlib-2.0.typelib
-r--r--r--   1 andrew  admin     640 Jan 23 11:17 xrandr-1.3.typelib
```
As I mentioned, scripts that don't use Cairo (and use other library) seem fine. For example, `gstplaystream.lua` plays great.

**Comment by psychon** [comment]
> I don't actually have that version installed I have 1.58.3 installed. If I adjust your command and make the link, it still doesn't find it (of course).

Hm... then create the necessary folders and create the link there? It's ugly, but it will surely work.

> Any clues on where the string '1.54.1' may be lingering?

When LGI is built, it gets a C define from.. something. And this is only used on MacOS... Anyway, this 1.54.1 comes from the C define `GOBJECT_INTROSPECTION_LIBDIR`. No idea where that comes from; it's not defined for me on Linux. LGI just takes that C define and embeds the name and version of the library that it wants to load to that.

Sorry, I don't know more than that.

**Comment by andrewstarks** [comment]
@psychon I cannot thank you enough for your help. I also understand that I'm effectively reducing you to tech support. Please know that I'm researching the problem as much as possible before I bug you.

I uninstalled everything, including LGI. I cleaned out homebrew (update, doctor, uninstalled everything, reinstalled, no errors)

After doing that, this happens:

```
andrew:~/ $ pkg-config --exists 'gobject-introspection-1.0 >= 0.10.8' --print-errors                                     
andrew:~/ $ pkg-config --exists libffi --print-errors                                                                    
andrew:~/ $ more $PKG_CONFIG_PATH/libffi.pc                                                                              
prefix=/usr/local/Cellar/libffi/3.2.1
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
toolexeclibdir=${exec_prefix}/lib
includedir=${libdir}/libffi-3.2.1/include

Name: libffi
Description: Library supporting Foreign Function Interfaces
Version: 3.2.1
Libs: -L${toolexeclibdir} -lffi
Cflags: -I${includedir}
andrew:~/ $ ls /usr/local/Cellar/libffi/3.2.1
total 384
drwxr-xr-x  9 andrew  staff     288 Jan 24 14:51 .
drwxr-xr-x  3 andrew  staff      96 Jan 24 14:51 ..
drwxr-xr-x  3 andrew  staff      96 Nov 12  2014 .brew
-rw-r--r--  1 andrew  staff  171152 Nov 12  2014 ChangeLog
-rw-r--r--  1 andrew  staff     606 Jan 24 14:51 INSTALL_RECEIPT.json
-rw-r--r--  1 andrew  staff    1132 Nov 12  2014 LICENSE
-rw-r--r--  1 andrew  staff   16155 Jan 24 14:51 README
drwxr-xr-x  7 andrew  staff     224 Nov 12  2014 lib
drwxr-xr-x  4 andrew  staff     128 Nov 12  2014 share
andrew:~/ $ sudo luarocks install lgi
Warning: falling back to curl - install luasec to get native HTTPS support
Installing https://luarocks.org/lgi-0.9.2-1.src.rock
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C lgi
pkg-config --exists 'gobject-introspection-1.0 >= 0.10.8' --print-errors
Package libffi was not found in the pkg-config search path.
Perhaps you should add the directory containing `libffi.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libffi', required by 'gobject-introspection-1.0', not found
make[1]: *** [.depcheck] Error 1
make: *** [all] Error 2

Error: Build error: Failed building.
```
Any help is appreciated. Thank you again.

**Comment by psychon** [comment]
> I also understand that I'm effectively reducing you to tech support

I am tech support here ;-)
(What does that say about LGI if LGI's tech support has no clue about MacOS, homebrew, ....?)

> andrew:~/ $ pkg-config --exists libffi --print-errors  
> [...]
> andrew:~/ $ sudo luarocks install lgi
> [...]
> Package libffi was not found in the pkg-config search path.

Uhm... `export PKG_CONFIG_PATH`? Actually, I guess that `sudo` somehow cleans up the environment for security reasons. `sudo sh -c 'pkg-config --exists libffi --print-errors && luarocks install lgi'` would test for that (and `sudo sh -c "PKG_CONFIG_PATH='$PKG_CONFIG_PATH' luarocks install lgi"` could Just Work (tm)).

**Comment by andrewstarks** [comment]
It compiled, which is nice (back to original problem)

```
andrew:~/ $ sudo sh -c 'pkg-config --exists libffi --print-errors && luarocks install lgi'                          
Password:
Package libffi was not found in the pkg-config search path.
Perhaps you should add the directory containing `libffi.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libffi' found
andrew:~/ $ sudo sh -c "PKG_CONFIG_PATH='$PKG_CONFIG_PATH' luarocks install lgi          
```
... then it compiled and installed.


Testing with Cairo...
```
andrew:gst_play/ $ lua gstvideo.lua                                                                     
lua: /usr/local/share/lua/5.3/lgi/override/cairo.lua:31: attempt to index a nil value (field '_module')
stack traceback:
	/usr/local/share/lua/5.3/lgi/override/cairo.lua:31: in main chunk
	[C]: in function 'require'
	/usr/local/share/lua/5.3/lgi/namespace.lua:183: in function 'lgi.namespace.require'
	/usr/local/share/lua/5.3/lgi/namespace.lua:170: in function 'lgi.namespace.require'
	/usr/local/share/lua/5.3/lgi/namespace.lua:170: in function 'lgi.namespace.require'
	/usr/local/share/lua/5.3/lgi/namespace.lua:170: in function 'lgi.namespace.require'
	(...tail calls...)
	gstvideo.lua:10: in main chunk
	[C]: in ?
andrew:gst_play/ $ lua -e 'print(require("lgi.core").module("cairo", 2))'                                           
nil	/usr/local/Cellar/gobject-introspection/1.58.3/lib/libcairo.2.dylib
```

sad panda.

**Comment by psychon** [comment]
Sorry, but I cannot tell you more than `sudo mkdir -p /usr/local/Cellar/gobject-introspection/1.58.3/lib/ && sudo ln -s /usr/local/Cellar/cairo/HEAD-ef8c379/lib/libcairo.2.dylib  /usr/local/Cellar/gobject-introspection/1.54.1/lib/libcairo.2.dylib`. And yes, this is ugly.

**Comment by andrewstarks** [comment]
I hear yeah. I tried that and it broke.

I'm interested in helping you fix the Mac experience with your library. Goal might be to make it so that you can get anyone with a relatively modern release OSX up and running without weirdness.  

Feel free to contact me directly if you're interested in exploring this. Thanks!

**Comment by rolandlo** [comment]
I thinks this issue can be closed now that #331 has been merged. There is no need to add any symlinks any more (with GLIB >= 2.76) on master.
